src/base/init/init.c

```diff
@@ -321,6 +321,29 @@ static void *mem_reserve(uint32_t memsize)
   return result;
 }
 
+static void low_mem_init_config_scrub(void)
+{
+  const uint32_t mem_1M = 1024 * 1024;
+  /* 16Mb limit is for being in reach of DMAc */
+  const uint32_t mem_16M = mem_1M * 16;
+
+  if (config.dpmi_base < LOWMEM_SIZE + HMASIZE + EXTMEM_SIZE + config.xms_size*1024) {
```

Comment:
I think xms_size should not be here.
It is not mapped unless you call lock_EMB().
So it can be of any size but lock_EMB()
will eventually fail. You need to use
xms_map_size here.

Reply:
xms_map_size isn't initialized yet here. So I can't use it.
it would simply be
`if (config.dpmi_base < LOWMEM_SIZE + HMASIZE + EXTMEM_SIZE)`
then.

So of course I can set config.xms much bigger than config.xms_map_size. But then I get the noisy crusader back.
Why can't a failing lock_EMB simply return an error code to the caller instead of also printing to the terminal screen, if it isn't fatal?

Reply:
> if (config.dpmi_base < LOWMEM_SIZE + HMASIZE + EXTMEM_SIZE)

You can write it as
`if (config.dpmi_base < LOWMEM_SIZE + HMASIZE + EXTMEM_SIZE + mem_1Mb)`
if we think that this is a minimum for
xms_map_size. Or maybe just write:
`if (config.dpmi_base < mem_16Mb)`
?

> Why can't a failing lock_EMB simply return an error code to the caller instead of also printing to the
> terminal screen, if it isn't fatal?

It certainly can, but just silencing the
warning is not sufficient. Currently it
doesn't cross the 16Mb and we only
guarantee 1Mb for committed space.
This is definitely too few for VCPI.

Reply:
VCPI can live with 1MB of committed space since clients can use VCPI (EMS) memory instead. I think the issue is really here that Crusader Demo really tries very hard to map all XMS memory, where the other program's I've tried just get the largest block (set to the max map size unless 0 left) and stop there.

So I think with VCPI it's really best to have all XMS memory available to lock; outside VCPI the present method is ok.

Taking this further I was thinking that perhaps things were better without the reverts of #1881 and would actually like that to be unreverted! Reason is that KVM can map that physical RAM as physical RAM, e.g. it can map 0x110000 directly to shm, or kvm_base could be extended, and then 0x110000 in KVM maps to &kvm_base[0x110000], which can alias map to shm for EMB locked mappings. Then, when DPMI kicks in 0x110000 can map again to &mem_base[0x110000] (I really don't want VCPI and dosemu-provided-DPMI at the same time!).

Reply:
> So I think with VCPI it's really best to have all XMS memory available to lock

Yes, that's what I am saying too.
So why not to do just that?

> actually like that to be unreverted!

That will mean non-identity mapping.
Which I think currently gives more
problems than profits.

> Reason is that KVM can map that physical RAM as physical RAM, e.g. it can map 0x110000 directly to shm,

Or directly to &mem_base[0x110000]?
What's the difference?

> or kvm_base could be extended

Well, yes, with current code it can't, but you
say "or". The first option still works?
We don't yet have a practice of extending
non-main mem_base, as well as we don't
have a big practice (aside from a few weeks
before #1881 reverts) to have non-identity
mappings.
The reason why I decided to move them away
from shm was to share ext_mem with dpmi_rsv_low,
reverting to identity mapping that way.
I still don't see the exact problem with that
approach, so why do you want that undone?

> Then, when DPMI kicks in 0x110000 can map again to &mem_base[0x110000]

Well, that's an interesting trick, but why
such a complexity? Note that DPMI doesn't
even need that region normally. It may use
it only for fixed mappings (aka HX), and I
don't think VCPI can use it at the same time?

Reply:
Oh, you seem to want kvm_base and mem_base
diverge, with only kvm flipping between them?
And only kvm_base having shm for ext_mem,
while mem_base has private mapping there?
I am a bit confused, you probably need to clarify
your idea a bit more. :) I don't think diverged
bases can work, at least because himem.sys
will then access only one of them.

Reply:
> (I really don't want VCPI and dosemu-provided-DPMI at the same time!).

But there is himem.sys which makes the
need for ext_mem to be available for everyone.
So I don't understand.

Reply:
I think himem would be ok, since before 1881 it uses shm (`ptr = alloc_mapping(MAPPING_EXTMEM, EXTMEM_SIZE);`). VCPI would just see that physically at 0x110000, but it's not mapped at &mem_base[0x110000], instead some higher linear address that DPMI can query (much like for the VGAEMU LFB).

to clarify what I'd like essentially is this memory layout for *physical* memory during VCPI execution:
```
0x000000 - low+HMA
0x110000 - internal XMS OR ext mem (int15/88 gives 0 with internal XMS active as before 1881)
0x110000+$_xms OR $_ext_mem - EMS/VCPI memory
0x110000+$_xms OR $_ext_mem + $_ems - top of RAM
```
but during DPMI execution `0x110000` would map to `&mem_base[0x110000]` which would have the same layout as before 1881.

Yes the non-indentity maps are a little more complex but then:
a) it avoids the hole for extmem when internal xms is used
b) himem can't map &mem_base[0x400000] which overlaps with HX
c) I don't need to worry about the 16MB or $_dpmi_base any more, as XMS/extmem physical RAM always starts at 0x110000 and no longer overlaps with dpmi_base.

Reply:
> a) it avoids the hole for extmem when internal xms is used

Why is that a problem if we remove
the current 16Mb limit? Its then the
same thing.

> b) himem can't map &mem_base[0x400000] which overlaps with HX

But it shouldn't because himem is not
a separate process. If HX doesn't ask it
to allocate an EMB, then it doesn't touch
any memory.

> c) I don't need to worry about the 16MB

But not because of moving himem to shm,
but rather because you want to disable
either xms or ext_mem. If you do only that,
you already do not worry about 16Mb too
much.

> starts at 0x110000 and no longer overlaps with dpmi_base.

But it doesn't do even now.
Probably you meant the rsv_low stuff?

So yes, this layout can be used. We know
that since I tried it already. :) But its much
more complex, and most of the advantages
you cited, are either because we currently
limit to 16Mb or because we currently enable
both ext_mem and xms together.
Please list the advantages specific to the
non-identity layout, if any.

Reply:
> But it shouldn't because himem is not
> a separate process. If HX doesn't ask it
> to allocate an EMB, then it doesn't touch
> any memory.
I still feel a bit uneasy about this. Some program could allocate an EMB at 0x400000, and then call HX, and there could be a conflict no? Or will there be an error message for HX?

Reply:
the other issues could indeed be avoided by disabling ext_mem if $_xms > 0, and by increasing $_dpmi_base when $_xms > 31M (or reintroducing $_dpmi_base=(auto)).

Hopefully I'm simply missing something about the conflict between DPMI linear mappings and identity-mapped XMS.

Reply:
> Some program could allocate an EMB at 0x400000

AFAIK there is no "allocate EMB fixed"
to alloc EMB at 0x400000. Some prog
may allocate 1-2 EMBs, but fill up 3Mb
of an XMS memory before calling another
prog? Very bad practice to drain some
mem before calling a child process.

> and then call HX,

Which means, this prog is comcom32?
Because usually you start some win32
prog from the shell.
So while I think you can write a test-case
that breaks, in practice its kinda impossible.

So I am not quite sure this needs to be
done. Of course it depends on how much
you want that to happen, but IMHO you
just need to remove the 16Mb limit - a few
lines of a change.

Reply:
> and by increasing $_dpmi_base when $_xms > 31M

I don't think xms is _ever_ used that
much. All I've seen is its use of a few MBs.
But if you want to code up "(auto)" - oh
well, please do.

Reply:
And let me point one more thing.
Yes, that non-linear layout looks
exceptionally cleaner. Its like oh
wow, I can move away all those
xms mappings, get rid of any
possible conflicts and limitations.
Which was why I implemented it
eventually.
But it all blows up when you realize
that the 16Mb DMA is not going
anywhere. That you still can't have
plenty of ext_mem+xms without one
of them being out of DMA space.
And if you end up enabling only
one of them, rather than both, then
you quickly realize that you could
do the same by starting xms mem
from 8Mb. You still have 8Mb for
DMA and any amount up to dpmi_base.
So again nothing cool comes out.
I went through all of that and just
concluded that its not worth a trouble.
You win only when you disable either
one, but then you don't need a non-identity
layout at all.

Reply:
So if you want to "solve all problems and
save the world", you'll start from having
only one of ext_mem/xms enabled. That
makes things a bit cleaner, and after that,
switching to non-identity layout doesn't
add anything at all.

Reply:
ok thanks for explaining. I was under the impression that 1881 was to specifically to help VCPI, and just saying it's not needed, but if you have other reasons to keep it simple, then ok.

I know the likelihood of conflicts (with HX) is extremely small, the question is then really is the extra complexity worth the trouble. From my point of view, I see no downsides other than the complexity, so if we can eliminate the conflict, then even if it's unlikely why not? But we can agree to disagree here, it won't impact any users I'm fairly sure.

Reply:
I adapted the patch in this PR to just do the bare minimum you requested (apart from a little more info in the error messages)

Reply:
anything else could just be bonus, but no time to do that properly now.

Reply:
Yes, that was to help VCPI (or kvm,
to keep the identity maps), but the other
reason was simple: it isn't helping.
Yes, it solves a fairly theoretical conflict,
but it doesn't allow growing xms or ext_mem,
which looks like a more real problem, esp
in a light of VCPI. And the other solution -
disabling either xms or vcpi - actually solves
things, rather than theoretically. So if you
really want to eliminate the conflict, you
can as well consider disabling either xms
or vcpi - then you are free of the DMA
problem and can increase xms to any size.
So the point is, getting to non-identity
layout is only meaningful under following conditions:
1 you want to solve the very theoretical conflict
2 you need both ext_mem and xms enabled
3 you agree on them being not growable
  above default values
4 you do not need the simplicity of identity mapping

If either of the above is not true, then you will
either disable one of xms/ext_mem, or do
nothing. Because disabling solves more problems
than non-identity layout, and adds no complexity.

Reply:
Also note that the conflict is only
possible if you drain xms of external
himem.sys, not internal xms.
But actually... if you really want
non-identity mapping, then at least
do not bind it to xms, as I did before.
Maybe you can get something simple
if you share everything up to dpmi_base,
and just alias it twice: to the beginning
of mem_base, and to the end?